### ISSUE TITLE
Optimize fmax with NAN

### ISSUE THREAD

**Issue by naoyam** [NVIDIA/Fuser#319]
Fp max reductions would typically look like:

```
  for(nvfuser_index_t i154 = 0; i154 < 8; ++i154) {
    int i299;
    i299 = 4 * i154;
#pragma unroll
      for(nvfuser_index_t i156 = 0; i156 < 4; ++i156) {
        T29[0] = fmax(
            T29[0],
            T24[(i299 + i156)]);
      }
  }
```

Here, fmax is is not just `fmaxf`, but it also incurs two more comparisons in case the arguments are NAN: https://github.com/NVIDIA/Fuser/blob/main/runtime/helpers.cu#LL102C1-L111C2

This could be translated as:

```
  bool is_nan = false;
#pragma unroll
  for(nvfuser_index_t i154 = 0; i154 < 8; ++i154) {
    int i299;
    i299 = 4 * i154;
#pragma unroll
      for(nvfuser_index_t i156 = 0; i156 < 4; ++i156) {
#if 0
        T29[0] = fmax(
            T29[0],
            T24[(i299 + i156)]);
#else
        T29[0] = T29[0] > T24[(i299 + i156)] ? T29[0] : T24[(i299 + i156)];
        is_nan = is_nan || isnan(T24[(i299 + i156)]);
#endif
      }
  }
  if (is_nan) {
    T29[0] = NAN;
  }
```

In the case of cross entropy loss (#278), I observed 20% speedup on A100.

I think this translation could be done automatically as part of lowering. See the translation for welford vectorization.

**Comment by jacobhinkle** [comment]
This is a good idea to speed up every call to `max` (and `min`). Furthermore in the particular case of softmax I think the nan checks could be avoided entirely since even if the max were non-nan the resulting softmax still will be nan due to the later sum, so propagation happens automatically. In that case we'd prefer a true `fmaxf`-style max computation with no nan check (like `torch.fmax` instead of `torch.maximum`). We could provide two different binary ops for each of min and max and add a `propagate_nan=true` option to the max and min functions, which we would set to false for `softmax` and `log_softmax`.

**Comment by jacobhinkle** [comment]
We could even determine whether nan checks can be skipped automatically without user input, but it involves a bit of a complicated traversal. We would visit each min or max op or reduction and mark the result as `unchecked` and the input as `propagated`, then move downstream from further uses of the input and from the min/max output, propagating the tags. When a pointwise binary or ternary op other than min or max has an `unchecked` input and a `propagated` input, it resolves the might-be-unchecked flag and becomes properly propagated. If we reach an output or another reduction we stop and if any output is not marked as propagated it means we can't skip nan checks in the original op.

**Comment by jacobhinkle** [comment]
For the immediate task at hand, I believe we can remove one nan check more easily by removing this branch: https://github.com/NVIDIA/Fuser/blob/main/runtime/helpers.cu#L106-L107. If `a == a`, then `a > b` will be false if `b != b` so the else branch subsumes the `b != b` branch. Similarly, for `fmin` we can remove the `a != a` branch. Doing this, we remove one nan check per reduction element, so it should be equivalent in speed to code in the description of this issue, but doesn't necessitate any codegen changes. As mentioned in my other comments, we can probably remove all the nan checks but I will leave that for a separate PR.

**Comment by jacobhinkle** [comment]
Interesting. Have you tried it without any nan checking? I am not sure how `fmaxf` works internally but it seems like it would need to branch since it guarantees to return non-nan in case only one input is nan. On the other hand, without nan checks i.e. just `a > b ? a : b` might return nan or not, we don't care!

**Comment by jacobhinkle** [comment]
Interesting! Maybe we should re-open and I'll get to work on the original approach.

**Comment by naoyam** [comment]
@kevinstephano mentioned this could be a good first issue.

**Comment by liqiangxl** [comment]
Can we make NaN propagation in fmax/fmin optional? Using `fmaxf` leads to 1.1x to 1.2x speedup for softmax on GB200, see #5065 ?

**Comment by liqiangxl** [comment]
> This is a good idea to speed up every call to `max` (and `min`). Furthermore in the particular case of softmax I think the nan checks could be avoided entirely since even if the max were non-nan the resulting softmax still will be nan due to the later sum, so propagation happens automatically. In that case we'd prefer a true `fmaxf`-style max computation with no nan check (like `torch.fmax` instead of `torch.maximum`). We could provide two different binary ops for each of min and max and add a `propagate_nan=true` option to the max and min functions, which we would set to false for `softmax` and `log_softmax`.

Sounds a good plan.

**Comment by naoyam** [comment]
> Can we make NaN propagation in fmax/fmin optional? Using `fmaxf` leads to 1.1x to 1.2x speedup for softmax on GB200, see [#5065](https://github.com/NVIDIA/Fuser/pull/5065) ?

No, I don't think we should.

**Comment by liqiangxl** [comment]
> > Can we make NaN propagation in fmax/fmin optional? Using `fmaxf` leads to 1.1x to 1.2x speedup for softmax on GB200, see [#5065](https://github.com/NVIDIA/Fuser/pull/5065) ?
> 
> No, I don't think we should.

what about skip nan check in `softmax` and `log_softmax`?

>  in the particular case of softmax I think the nan checks could be avoided entirely since even if the max were non-nan the resulting softmax still will be nan due to the later sum, so propagation happens automatically. In that case we'd prefer a true fmaxf-style max computation with no nan check

**Comment by naoyam** [comment]
I'm yet convinced why the simpler approach I mentioned at the beginning is not enough. At least on A100, that was close to optimal. Jacob's analysis is interesting, but it only works when two reductions are executed sequentially like softmax.

**Comment by liqiangxl** [comment]
> I'm yet convinced why the simpler approach I mentioned at the beginning is not enough. At least on A100, that was close to optimal. Jacob's analysis is interesting, but it only works when two reductions are executed sequentially like softmax.

I feel using the intrinsic `fmaxf` is much simpler than modifying nvFuser to generate the code using `isnan()`
I manually checked the `isnan()` approach on GB200, it leads to lower performance of [cross entropy benchmark](https://github.com/NVIDIA/Fuser/blob/119d6b2c7a88fe2e5dc2f60cbdefa2f4adb666ef/benchmarks/python/test_cross_entropy_loss.py#L129).
(1) using `fmaxf`: 60% SOL
(2) using `fmax`: 50%  SOL
(3) using 'isnan': 48% SOL (still use fmax in warp reduction)

<img width="254" height="334" alt="Image" src="https://github.com/user-attachments/assets/043f029e-17e9-4b05-8912-640f811b8622" />

Because of `unroll`, these `isnan` checks are explicitly executed..

<img width="1049" height="418" alt="Image" src="https://github.com/user-attachments/assets/141f2fb8-79d0-4270-ac7e-4d67fa9170b3" />

**Comment by liqiangxl** [comment]
> likely because the root cause of the overhead would be the nan-check branch, which is almost completely eliminated.

If there are no NANs in the input (I assume NaN is rare),`is_nan == false`, none of these checks is eliminated.

**Comment by naoyam** [comment]
> none of these checks is eliminated.

I meant the branch for handling the nan. In your example, the branch is executed 128 times less than the original since the branch is only executed once after the nested loops. That's why said it's almost eliminated.
